File: java/iterator/EmployeeCoWorkerIterator.java
package iterator;

public class EmployeeCoWorkerIterator implements EmployeeHierarchyIterator {
    private final EmployeeHierarchyCollection employeeHierarchyCollection;
    private int currentCoWorkerPosition = 0;

    public EmployeeCoWorkerIterator(EmployeeHierarchyCollection employeeHierarchyCollection) {
        this.employeeHierarchyCollection = employeeHierarchyCollection;
    }

    @Override
    public boolean hasNext() {
        return currentCoWorkerPosition < employeeHierarchyCollection.getEmployees().size();
    }

    @Override
    public Employee getNext() {
        System.out.println("Iterating through a graph DFS / BFS");
        if (!hasNext()) {
            return null;
        }
        return employeeHierarchyCollection.getEmployees().get(currentCoWorkerPosition++);
    }
}

File: typescript/patterns/iterator/employee-coworker-iterator.ts
class EmployeeCoWorkerIterator implements EmployeeHierarchyIterator {
  private employeeHierarchyCollection: EmployeeHierarchyCollection;
  private currentCoWorkerPosition: number = 0;

  constructor(employeeHierarchyCollection: EmployeeHierarchyCollection) {
    this.employeeHierarchyCollection = employeeHierarchyCollection;
  }

  getNext(): Employee | null {
    console.log("Iterating through a graph DFS / BFS");
    if (!this.hasNext()) return null;
    return this.employeeHierarchyCollection.getEmployees()[
      this.currentCoWorkerPosition++
    ];
  }

  hasNext(): boolean {
    return (
      this.currentCoWorkerPosition <
      this.employeeHierarchyCollection.getEmployees().length
    );
  }
}
